File: 17/py/solve.py
#!/usr/bin/env python3

import sys


class Spinlock():
    def __init__(self, stride):
        self._vec = [0]
        self._pos = 0
        self._stride = stride + 1

    def insert(self, val):
        self._vec.insert(self._next_pos(), val)

    def get_after(self):
            return self._vec[(self._pos + 1) % len(self._vec)]

    def _next_pos(self):
        self._pos += self._stride
        self._pos %= len(self._vec)

        return self._pos


def solve_1(spinlock, max_val):
    for i in range(1, max_val+1):
        spinlock.insert(i)

    return spinlock.get_after()


def main():
    input = 348
    spinlock = Spinlock(input)

    print("Answer 1: %d" % solve_1(spinlock, 2017))

    return True


if __name__ == '__main__':
    sys.exit(not main())

File: 17/cpp/solve.cpp
#include <iostream>


class List_item
{
    public:
        List_item(int val, List_item *next): val(val), next(next) {}

        List_item *set_next(List_item *iNext) {next = iNext; return this;}

        List_item *get_next() {return next;}

        int get_val() {return val;}

    private:
        int val;
        List_item *next;
};


class Spinlock
{
    public:
        Spinlock(size_t stride):
            stride(stride)
        {
            pos = new List_item(0, NULL);
            pos->set_next(pos);
        }

        void insert(int val)
        {
            auto parent = next_pos();
            auto successor = parent->get_next();
            parent->set_next(new List_item(val, successor));
            pos = parent->get_next();
        }

        int get_after_val(int val) const
        {
            auto tmp_pos = pos;
            while (true)
            {
                if (tmp_pos->get_val() == val)
                {
                    return tmp_pos->get_next()->get_val();
                }
                else
                {
                    tmp_pos = tmp_pos->get_next();
                }
            }
        }

    private:
        List_item *next_pos()
        {
            for (size_t i=0; i<stride; i++)
            {
                pos = pos->get_next();
            }

            return pos;
        }


        List_item *pos;
        size_t stride;
};


int main(void)
{
    size_t stride = 348;
    size_t max_v = 50000000;
    Spinlock spinlock(stride);

    for (size_t i=1; i<=max_v; i++)
    {
        spinlock.insert(i);

        if (i%100000 == 0)
        {
            std::cout << i/1000 << '\n';
        }
    }

    std::cout << "Answer to part 2 is: " << spinlock.get_after_val(0) << '\n';

    return 0;
}

